File: UpdateZentrale/Providers/RepoSkriptAktualisierer.cs
using System.IO;
using System.Text.RegularExpressions;
using UpdateZentrale.Models;
using UpdateZentrale.Services;

namespace UpdateZentrale.Providers;

/// <summary>
/// Own tools in ~/proggs. Updating never reimplements the existing PowerShell scripts, it only
/// calls them -- and never with -Force, because each script opens its own yes/no window that the
/// user has to answer.
/// </summary>
public sealed class RepoSkriptAktualisierer : IAktualisierer
{
    public string Art => "reposkript";

    public async Task<PruefErgebnis> PruefenAsync(ProgrammEintrag eintrag, IProgress<string> protokoll, CancellationToken abbruch)
    {
        var installiert = InstallierteVersion(eintrag);
        var quelle = QuellVersion(eintrag);

        // Signal 1: does the remote carry commits for this folder's build sources that are not here yet?
        var hinterstand = await HinterstandAsync(Pfade.RepoWurzel, eintrag.RepoOrdner, protokoll, abbruch);

        // Signal 2: is the source version newer than the built exe? Compared numerically --
        // a prefix test called "1.24.5" current for an exe built from "1.24.50".
        var quellstandNeuer = !string.IsNullOrWhiteSpace(quelle)
                              && !string.IsNullOrWhiteSpace(installiert)
                              && CliAktualisierer.Vergleiche(quelle, installiert) > 0;

        protokoll.Report("gebaut: " + installiert + " | Quelle: " + quelle);

        return Einordnen(installiert, quelle, hinterstand, quellstandNeuer, !string.IsNullOrWhiteSpace(eintrag.RepoOrdner));
    }

    /// <summary>
    /// The verdict of a check. A git query that failed is "unknown", never "current": the card
    /// then keeps the update button instead of hiding a pending build behind a green pill.
    /// </summary>
    internal static PruefErgebnis Einordnen(string installiert, string quelle, int? hinterstand,
                                            bool quellstandNeuer, bool hatRepoOrdner)
    {
        if (string.IsNullOrWhiteSpace(installiert))
            return new PruefErgebnis(UpdateZustand.NichtInstalliert, "", quelle, "Noch nicht gebaut.");

        if (hinterstand > 0 || quellstandNeuer)
        {
            var grund = hinterstand > 0
                ? hinterstand + " neue Commit(s) mit Quellcode auf origin/main"
                : "Quellstand " + quelle + " neuer als der gebaute Stand";
            return new PruefErgebnis(UpdateZustand.UpdateVerfuegbar, installiert, quelle, grund + ".");
        }

        if (hatRepoOrdner && hinterstand is null)
            return new PruefErgebnis(UpdateZustand.Unbekannt, installiert, quelle,
                "Der Git-Stand ließ sich nicht abfragen – ob neue Commits vorliegen, ist unbekannt.");

        return new PruefErgebnis(UpdateZustand.Aktuell, installiert, quelle, "Gebauter Stand ist aktuell.");
    }

    public async Task<PruefErgebnis> AktualisierenAsync(ProgrammEintrag eintrag, IProgress<string> protokoll, CancellationToken abbruch)
    {
        var skript = Path.Combine(Pfade.RepoWurzel, (eintrag.Skript ?? "").Replace('/', Path.DirectorySeparatorChar));
        if (!File.Exists(skript))
            return new PruefErgebnis(UpdateZustand.Fehler, Meldung: "Skript fehlt: " + skript);

        // The check reports "update available" as soon as origin/main carries commits for this
        // folder, but the scripts only build what is on disk. Without pulling first they answer
        // "already current", the next check finds the same commits again, and the card asks for
        // the same update over and over. Same fast-forward-only pull the launcher itself does.
        var hinterstand = await HinterstandAsync(Pfade.RepoWurzel, eintrag.RepoOrdner, protokoll, abbruch);
        var lokalNeuer = LokaleQuelleNeuer(eintrag, protokoll);
        if (!DarfSkriptStarten(hinterstand, !string.IsNullOrWhiteSpace(eintrag.RepoOrdner), lokalNeuer))
            return new PruefErgebnis(UpdateZustand.Fehler,
                Meldung: "Der Git-Stand ließ sich nicht abfragen, und lokal liegt kein neuerer Quellcode als im "
                         + "gebauten Programm. Das Update-Skript wurde nicht gestartet – es würde nur den "
                         + "unveränderten Stand bauen. Netz und Repo prüfen, dann erneut versuchen.");
        if (hinterstand is null && !string.IsNullOrWhiteSpace(eintrag.RepoOrdner))
            protokoll.Report("Git-Stand unbekannt, aber lokal liegt neuerer Quellcode – das Skript baut den vorhandenen Checkout.");
        if (hinterstand > 0)
        {
            protokoll.Report("git pull --ff-only (" + hinterstand + " neue Commit(s) für " + eintrag.RepoOrdner + ")");
            var zug = await Kommandozeile.AusfuehrenAsync("git", "pull --ff-only", TimeSpan.FromMinutes(3), Pfade.RepoWurzel, abbruch);
            protokoll.Report(zug.Ausgabe);
            if (zug.ExitCode != 0)
                return new PruefErgebnis(UpdateZustand.Fehler,
                    Meldung: "git pull --ff-only ist fehlgeschlagen (lokale Änderungen oder abweichender Verlauf) – "
                             + "das Repo muss erst von Hand abgeglichen werden. Das Update-Skript wurde nicht gestartet.",
                    Protokoll: zug.Ausgabe);
        }

        var args = ("-NoProfile -ExecutionPolicy Bypass -File \"" + skript + "\" " + eintrag.SkriptArgumente).TrimEnd();
        protokoll.Report("pwsh " + args);
        protokoll.Report("Das Skript zeigt jetzt ein Ja/Nein-Fenster. Ohne Klick auf 'Ja' passiert nichts.");

        // The scripts wait up to 240 s for the click, so the timeout must sit clearly above that.
        var zeitlimit = TimeSpan.FromSeconds(Math.Max(eintrag.DialogWartezeitSekunden, 300) + eintrag.ZeitlimitMinuten * 60);
        var exeVorher = await FingerabdruckAsync(eintrag, abbruch);
        var lauf = await Kommandozeile.AusfuehrenAsync(PwshPfad(), args, zeitlimit, Pfade.RepoWurzel, abbruch);
        protokoll.Report(lauf.Ausgabe);
        var exeNachher = await FingerabdruckAsync(eintrag, abbruch);

        return Auswerten(lauf, eintrag.StatusPraefix, !string.IsNullOrWhiteSpace(exeVorher) && exeVorher == exeNachher);
    }

    /// <summary>
    /// Without a git answer the script would build whatever is on disk. That is only worth a run
    /// when the disk demonstrably holds newer build sources than the built exe -- otherwise it
    /// reproduces exactly the "same update again" loop.
    /// </summary>
    internal static bool DarfSkriptStarten(int? hinterstand, bool hatRepoOrdner, bool lokaleQuelleNeuer)
        => !hatRepoOrdner || hinterstand is not null || lokaleQuelleNeuer;

    /// <summary>
    /// Independent of git: is the source version newer than the exe, or is any build source file
    /// newer than it? The second test is the one update-launcher.ps1 itself uses.
    /// </summary>
    private static bool LokaleQuelleNeuer(ProgrammEintrag eintrag, IProgress<string> protokoll)
    {
        var exe = Pfade.Aufloesen(eintrag.ExePfadWirksam);
        var installiert = InstallierteVersion(eintrag);
        var quelle = QuellVersion(eintrag);
        if (!string.IsNullOrWhiteSpace(quelle) && !string.IsNullOrWhiteSpace(installiert)
            && CliAktualisierer.Vergleiche(quelle, installiert) > 0)
            return true;

        if (string.IsNullOrWhiteSpace(exe) || !File.Exists(exe) || string.IsNullOrWhiteSpace(eintrag.RepoOrdner))
            return false;
        try
        {
            var ordner = Path.Combine(Pfade.RepoWurzel, eintrag.RepoOrdner.Replace('/', Path.DirectorySeparatorChar));
            return QuelleNeuerAlsBuild(ordner, File.GetLastWriteTimeUtc(exe));
        }
        catch (Exception ex)
        {
            protokoll.Report("Lokaler Quellstand nicht lesbar: " + ex.Message);
            return false;
        }
    }

    /// <summary>True if a build source below the folder (bin/obj excluded) is newer than the build.</summary>
    internal static bool QuelleNeuerAlsBuild(string ordner, DateTime buildZeitUtc)
    {
        if (!Directory.Exists(ordner)) return false;
        var trenner = Path.DirectorySeparatorChar;
        return BauQuellen
            .SelectMany(m => Directory.EnumerateFiles(ordner, m, SearchOption.AllDirectories))
            .Where(d => !d.Contains(trenner + "bin" + trenner, StringComparison.OrdinalIgnoreCase)
                        && !d.Contains(trenner + "obj" + trenner, StringComparison.OrdinalIgnoreCase))
            .Any(d => File.GetLastWriteTimeUtc(d) > buildZeitUtc);
    }

    /// <summary>
    /// Maps the script's status line onto a card state. "started" alone is not proof of an update:
    /// update-launcher.ps1 also says "started" when the build was already current and it only
    /// launched the existing exe. Reported as a finished update, the unchanged fingerprint then
    /// turned it into "not verified", the card offered the same update again -- the endless loop.
    /// </summary>
    internal static PruefErgebnis Auswerten(BefehlErgebnis lauf, string? statusPraefix, bool buildUnveraendert)
    {
        // How the process ended outranks what it printed: a "started" written before a timeout,
        // a cancel or a tree that could not be ended fully is never a finished update.
        if (Kommandozeile.UnsauberesEnde(lauf) is { } unsauber) return unsauber;

        var status = StatusLesen(lauf.Ausgabe, statusPraefix);
        if (status == "cancelled")
            return new PruefErgebnis(UpdateZustand.Abgebrochen,
                Meldung: "Du hast im Ja/Nein-Fenster auf 'Nein' geklickt – es wurde nichts geändert.", Protokoll: lauf.Ausgabe);
        if (status == "no-answer")
            return new PruefErgebnis(UpdateZustand.Abgebrochen,
                Meldung: "Kein Klick im Zeitfenster – es wurde nichts geändert.", Protokoll: lauf.Ausgabe);
        if (status == "already-current")
            return new PruefErgebnis(UpdateZustand.Aktuell, Meldung: "War bereits aktuell.", Protokoll: lauf.Ausgabe);
        if (status == "started")
            return buildUnveraendert
                ? new PruefErgebnis(UpdateZustand.Aktuell,
                    Meldung: "Der Build war bereits aktuell – das Skript hat ihn nur gestartet.", Protokoll: lauf.Ausgabe)
                : new PruefErgebnis(UpdateZustand.Fertig, Meldung: "Neue Version gebaut und gestartet.", Protokoll: lauf.Ausgabe);
        if (lauf.ExitCode == 0 && string.IsNullOrWhiteSpace(statusPraefix))
            return new PruefErgebnis(UpdateZustand.Fertig, Meldung: "Skript erfolgreich durchgelaufen.", Protokoll: lauf.Ausgabe);
        if (lauf.ExitCode == 0)
            // The script promises a status line; without it the outcome is unknown, whatever the
            // exit code says. The fingerprint comparison afterwards decides, never this guess.
            return new PruefErgebnis(UpdateZustand.Fertig,
                Meldung: "Skript endete ohne Statuszeile – das Ergebnis wird am Build-Stand geprüft.", Protokoll: lauf.Ausgabe);

        return new PruefErgebnis(UpdateZustand.Fehler,
            Meldung: "Das Skript endete mit Code " + lauf.ExitCode + ".", Protokoll: lauf.Ausgabe);
    }

    /// <summary>
    /// The csproj version only moves when someone bumps it, so it alone would report a false
    /// failure after a rebuild. The write time of the built exe is what actually changes here.
    /// </summary>
    public Task<string> FingerabdruckAsync(ProgrammEintrag eintrag, CancellationToken abbruch)
    {
        var exe = Pfade.Aufloesen(eintrag.ExePfadWirksam);
        if (string.IsNullOrWhiteSpace(exe) || !File.Exists(exe)) return Task.FromResult("");

        try
        {
            var zeit = File.GetLastWriteTimeUtc(exe).ToString("yyyy-MM-dd HH:mm:ss");
            var version = InstallierteVersion(eintrag);
            return Task.FromResult((string.IsNullOrWhiteSpace(version) ? "" : version + " vom ") + zeit);
        }
        catch (Exception diagAusnahme)
        {
            Diagnose.Gefangen(diagAusnahme, "provider", Schwere.Debug);
            return Task.FromResult("");
        }
    }

    /// <summary>
    /// Only files that end up in the build count. The folders also carry runtime and profile data
    /// (OpenLauncher/models.json, Profiles/**) that other machines commit all day; counted as
    /// "new commits", they announced an update the script then rightly refused to build -- and
    /// the card asked for the same update again. Same file kinds update-launcher.ps1 checks
    /// against the exe, plus the MSBuild companions that change a build just as much.
    /// </summary>
    internal static readonly string[] BauQuellen =
        { "*.cs", "*.xaml", "*.csproj", "*.props", "*.targets", "*.resx", "*.manifest" };

    /// <returns>
    /// Commits on origin/main that touch build sources of the folder; 0 without a folder; null
    /// when git could not answer -- never a silent 0, which would read as "current".
    /// </returns>
    internal static async Task<int?> HinterstandAsync(string repoWurzel, string? ordner,
                                                      IProgress<string>? protokoll, CancellationToken abbruch)
    {
        if (string.IsNullOrWhiteSpace(ordner)) return 0;

        var holen = await Kommandozeile.AusfuehrenAsync("git", "fetch --quiet", TimeSpan.FromMinutes(3), repoWurzel, abbruch);
        if (holen.Abgelaufen || holen.ExitCode != 0)
        {
            // A stale origin/main can answer "0 behind" while the real remote is ahead -- that
            // would read as "current". Without a fresh fetch the state is unknown, full stop.
            protokoll?.Report("git fetch fehlgeschlagen (" + (holen.Abgelaufen ? "Zeitlimit" : "Code " + holen.ExitCode)
                              + ") – Remote-Stand unbekannt: " + holen.Ausgabe);
            return null;
        }

        var muster = string.Join(" ", BauQuellen.Select(m => "\":(glob)" + ordner.Trim('/', '\\') + "/**/" + m + "\""));
        var zaehl = await Kommandozeile.AusfuehrenAsync("git",
            "rev-list --count HEAD..origin/main -- " + muster,
            TimeSpan.FromMinutes(2), repoWurzel, abbruch);

        int? anzahl = !zaehl.Abgelaufen && zaehl.ExitCode == 0 && int.TryParse(zaehl.Ausgabe.Trim(), out var z) ? z : null;
        protokoll?.Report("git rev-list HEAD..origin/main -- " + ordner + " (Quellcode) -> "
                          + (anzahl?.ToString() ?? "FEHLER Code " + zaehl.ExitCode + ": " + zaehl.Ausgabe));
        return anzahl;
    }

    /// <summary>rebuild-overlay.ps1 requires PowerShell 7; fall back only if pwsh is missing.</summary>
    private static string PwshPfad()
    {
        var kandidaten = new[]
        {
            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles), "PowerShell", "7", "pwsh.exe"),
            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Microsoft", "WindowsApps", "pwsh.exe")
        };
        return kandidaten.FirstOrDefault(File.Exists) ?? "powershell.exe";
    }

    internal static string StatusLesen(string ausgabe, string? praefix)
    {
        if (string.IsNullOrWhiteSpace(praefix)) return "";
        foreach (var zeile in ausgabe.Split('\n'))
        {
            var index = zeile.IndexOf(praefix, StringComparison.OrdinalIgnoreCase);
            if (index < 0) continue;
            var rest = zeile.Substring(index + praefix.Length).Trim();
            return rest.Split(' ', StringSplitOptions.RemoveEmptyEntries).FirstOrDefault() ?? "";
        }
        return "";
    }

    private static string InstallierteVersion(ProgrammEintrag eintrag)
    {
        var exe = Pfade.Aufloesen(eintrag.ExePfadWirksam);
        if (string.IsNullOrWhiteSpace(exe) || !File.Exists(exe)) return "";
        try
        {
            var info = System.Diagnostics.FileVersionInfo.GetVersionInfo(exe);
            return info.FileVersion ?? info.ProductVersion ?? "";
        }
        catch (Exception diagAusnahme)
        {
            Diagnose.Gefangen(diagAusnahme, "provider", Schwere.Debug);
            return "";
        }
    }

    /// <summary>Reads the csproj Version element -- the single source of truth for these tools.</summary>
    private static string QuellVersion(ProgrammEintrag eintrag)
    {
        if (string.IsNullOrWhiteSpace(eintrag.ProjektDatei)) return "";
        var datei = Path.Combine(Pfade.RepoWurzel, eintrag.ProjektDatei.Replace('/', Path.DirectorySeparatorChar));
        if (!File.Exists(datei)) return "";
        try
        {
            var treffer = Regex.Match(File.ReadAllText(datei), @"<Version>\s*([^<\s]+)\s*</Version>");
            return treffer.Success ? treffer.Groups[1].Value : "";
        }
        catch (Exception diagAusnahme)
        {
            Diagnose.Gefangen(diagAusnahme, "provider", Schwere.Debug);
            return "";
        }
    }
}

File: UpdaterZentrale-macOS/UpdaterZentrale/Providers/RepoSkriptAktualisierer.swift
import Foundation

/// Eigene Werkzeuge in ~/proggs. 1:1-Port von Providers/RepoSkriptAktualisierer.cs.
///
/// Das Aktualisieren baut die vorhandenen Skripte NIE nach, es ruft sie nur auf -- und niemals mit
/// einem Force-Schalter, weil jedes Skript seine eigene Rueckfrage stellen soll.
///
/// Zwei Unterschiede zur Windows-Fassung:
///
/// 1. **`bash` statt `pwsh`**, und die Skripte heissen anders: update-launcher.sh statt
///    update-launcher.ps1, rebuild-overlay.sh statt rebuild-overlay.ps1.
///
/// 2. **`rebuild-overlay.sh` hat auf dem Mac KEIN Ja/Nein-Fenster und gibt KEIN
///    `OVERLAY_UPDATE_STATUS=` aus** -- anders als seine Windows-Schwester. Es beendet die
///    Overlays und baut sofort. Damit trotzdem nichts ungefragt passiert, stellt die App die
///    Rueckfrage bei solchen Eintraegen (leeres `statusPraefix`) SELBST, bevor das Skript
///    ueberhaupt startet. Das Ergebnis wird dann aus Exit-Code und Schlusszeile gelesen.
///    Das Skript selbst bleibt unangetastet: es gehoert einem anderen Projekt.
struct RepoSkriptAktualisierer: Aktualisierer {
    let art = "reposkript"

    // MARK: - Pruefen

    func pruefen(_ eintrag: ProgrammEintrag, _ protokoll: Fortschritt) async -> PruefErgebnis {
        let installiert = installierteVersion(eintrag)
        let quelle = quellVersion(eintrag)

        // Merkmal 1: liegen auf der Gegenstelle Commits fuer diesen Ordner, die hier fehlen?
        var hinterstand = 0
        if let repoOrdner = eintrag.repoOrdner, !repoOrdner.istLeer {
            _ = await Kommandozeile.ausfuehren("/usr/bin/git", ["fetch", "--quiet"],
                                               zeitlimit: 180, arbeitsverzeichnis: Pfade.repoWurzel)
            let zaehl = await Kommandozeile.ausfuehren(
                "/usr/bin/git",
                ["rev-list", "--count", "HEAD..origin/main", "--", repoOrdner],
                zeitlimit: 120, arbeitsverzeichnis: Pfade.repoWurzel)
            hinterstand = Int(zaehl.ausgabe.trimmed) ?? 0
            protokoll.berichte("git rev-list HEAD..origin/main -- \(repoOrdner) -> \(hinterstand)")
        }

        // Merkmal 2: ist der Quellstand neuer als die gebaute Fassung?
        let quellstandNeuer = !quelle.istLeer && !installiert.istLeer
                              && !installiert.hasPrefix(quelle)

        protokoll.berichte("gebaut: \(installiert) | Quelle: \(quelle)")

        if installiert.istLeer {
            return PruefErgebnis(zustand: .nichtInstalliert, verfuegbareVersion: quelle,
                                 meldung: "Noch nicht gebaut.")
        }

        if hinterstand > 0 || quellstandNeuer {
            let grund = hinterstand > 0
                ? "\(hinterstand) neue Commit(s) auf origin/main"
                : "Quellstand \(quelle) neuer als der gebaute Stand"
            return PruefErgebnis(zustand: .updateVerfuegbar, installierteVersion: installiert,
                                 verfuegbareVersion: quelle, meldung: grund + ".")
        }

        return PruefErgebnis(zustand: .aktuell, installierteVersion: installiert,
                             verfuegbareVersion: quelle, meldung: "Gebauter Stand ist aktuell.")
    }

    // MARK: - Aktualisieren

    func aktualisieren(_ eintrag: ProgrammEintrag, _ protokoll: Fortschritt) async -> PruefErgebnis {
        let skript = (Pfade.repoWurzel as NSString).appendingPathComponent(eintrag.skript ?? "")
        guard FileManager.default.fileExists(atPath: skript) else {
            return PruefErgebnis(zustand: .fehler, meldung: "Skript fehlt: \(skript)")
        }

        let hatEigeneRueckfrage = !(eintrag.statusPraefix?.istLeer ?? true)

        // Skripte ohne eigenes Ja/Nein-Fenster: hier wird gefragt, sonst gar nicht.
        if !hatEigeneRueckfrage {
            let frage = "\(eintrag.name) wird jetzt neu gebaut und neu gestartet.\n\n"
                      + "Das Skript beendet das laufende Programm dafür selbstständig; nicht "
                      + "gespeicherte Eingaben gehen verloren. Der Vorgang dauert ein bis zwei Minuten.\n\n"
                      + "Jetzt aktualisieren?"
            let freigegeben = await MainActor.run { Dialoge.fragen(frage, "\(eintrag.name) neu bauen?") }
            if !freigegeben {
                return PruefErgebnis(zustand: .abgebrochen,
                                     meldung: "Du hast abgelehnt – es wurde nichts geändert.")
            }
            protokoll.berichte("Update per Klick freigegeben.")
        }

        var argumente = [skript]
        if let skriptArgumente = eintrag.skriptArgumente, !skriptArgumente.istLeer {
            argumente.append(contentsOf: Kommandozeile.zerlegen(skriptArgumente))
        }

        protokoll.berichte("bash " + argumente.joined(separator: " "))
        if hatEigeneRueckfrage {
            protokoll.berichte("Das Skript zeigt jetzt ein Ja/Nein-Fenster. Ohne Klick auf 'Ja' passiert nichts.")
        }

        // Die Skripte mit Dialog warten bis zu 240 s auf den Klick; das Zeitlimit muss klar
        // darueber liegen. Die ohne Dialog warten bis zu 300 s auf eine laufende Aufnahme.
        let zeitlimit = TimeInterval(max(eintrag.dialogWartezeitSekunden, 300) + eintrag.zeitlimitMinuten * 60)
        let lauf = await Kommandozeile.ausfuehren("/bin/bash", argumente,
                                                  zeitlimit: zeitlimit,
                                                  arbeitsverzeichnis: Pfade.repoWurzel)
        protokoll.berichte(lauf.ausgabe)

        // Weg A: das Skript meldet seinen Status selbst (update-launcher.sh).
        if hatEigeneRueckfrage {
            let status = statusLesen(lauf.ausgabe, eintrag.statusPraefix)
            switch status {
            case "cancelled":
                return PruefErgebnis(zustand: .abgebrochen,
                                     meldung: "Du hast im Ja/Nein-Fenster auf 'Nein' geklickt – es wurde nichts geändert.",
                                     protokoll: lauf.ausgabe)
            case "no-answer":
                return PruefErgebnis(zustand: .abgebrochen,
                                     meldung: "Kein Klick im Zeitfenster – es wurde nichts geändert.",
                                     protokoll: lauf.ausgabe)
            case "already-current":
                return PruefErgebnis(zustand: .aktuell, meldung: "War bereits aktuell.",
                                     protokoll: lauf.ausgabe)
            case "started":
                return PruefErgebnis(zustand: .fertig, meldung: "Neue Version gebaut und gestartet.",
                                     protokoll: lauf.ausgabe)
            default:
                break
            }
        }

        if lauf.abgelaufen {
            return PruefErgebnis(zustand: .fehler, meldung: "Zeitlimit überschritten.", protokoll: lauf.ausgabe)
        }

        // Weg B: kein Statusband -- Exit-Code und Schlusszeile entscheiden (rebuild-overlay.sh).
        if lauf.exitCode == 0 {
            let meldung = lauf.ausgabe.contains("Fertig — alle Ziele")
                       || lauf.ausgabe.contains("Fertig - alle Ziele")
                ? "Neue Version gebaut, gestartet und verifiziert."
                : "Skript erfolgreich durchgelaufen."
            return PruefErgebnis(zustand: .fertig, meldung: meldung, protokoll: lauf.ausgabe)
        }

        if lauf.ausgabe.contains("Fertig mit Problemen bei") {
            return PruefErgebnis(zustand: .fehler,
                                 meldung: "Das Skript konnte den Neustart nicht verifizieren – Einzelheiten im Protokoll.",
                                 protokoll: lauf.ausgabe)
        }

        return PruefErgebnis(zustand: .fehler, meldung: "Das Skript endete mit Code \(lauf.exitCode).",
                             protokoll: lauf.ausgabe)
    }

    // MARK: - Fingerabdruck

    /// Die Versionsnummer allein bewegt sich nur, wenn jemand sie hochzaehlt -- sie wuerde nach
    /// einem Neubau faelschlich einen Fehlschlag melden. Was sich hier wirklich aendert, ist die
    /// Schreibzeit der gebauten Programmdatei.
    func fingerabdruck(_ eintrag: ProgrammEintrag) async -> String {
        let exe = Pfade.aufloesen(eintrag.exePfad)
        guard !exe.isEmpty, FileManager.default.fileExists(atPath: exe) else { return "" }

        let zeit = Pfade.schreibZeit(exe)
        guard !zeit.isEmpty else { return "" }
        let version = installierteVersion(eintrag)
        return (version.istLeer ? "" : version + " vom ") + zeit
    }

    // MARK: - Intern

    private func statusLesen(_ ausgabe: String, _ praefix: String?) -> String {
        guard let praefix, !praefix.istLeer else { return "" }
        for zeile in ausgabe.split(separator: "\n", omittingEmptySubsequences: false) {
            let text = String(zeile)
            guard let stelle = text.range(of: praefix, options: .caseInsensitive) else { continue }
            let rest = String(text[stelle.upperBound...]).trimmed
            return rest.split(separator: " ").first.map(String.init) ?? ""
        }
        return ""
    }

    /// Die gebaute Fassung. Unter Windows war das FileVersionInfo der exe; auf dem Mac liegt die
    /// Version in der Info.plist des gebauten App-Buendels.
    private func installierteVersion(_ eintrag: ProgrammEintrag) -> String {
        // Bevorzugt das App-Buendel; sonst das Buendel um die Programmdatei herum ableiten
        // (<App>.app/Contents/MacOS/<Name> -> <App>.app).
        if let appPfad = eintrag.appPfad, !appPfad.istLeer {
            let pfad = Pfade.aufloesen(appPfad)
            if FileManager.default.fileExists(atPath: pfad) { return Pfade.appVersion(pfad) }
        }

        let exe = Pfade.aufloesen(eintrag.exePfad)
        guard !exe.isEmpty, FileManager.default.fileExists(atPath: exe) else { return "" }

        let macOsOrdner = (exe as NSString).deletingLastPathComponent          // …/Contents/MacOS
        let contents = (macOsOrdner as NSString).deletingLastPathComponent     // …/Contents
        let buendel = (contents as NSString).deletingLastPathComponent         // …/<App>.app
        guard buendel.hasSuffix(".app") else { return "" }
        return Pfade.appVersion(buendel)
    }

    /// Der Soll-Stand aus der Quelle. Unter Windows das `<Version>`-Element der csproj, auf dem
    /// Mac das CFBundleShortVersionString der Info.plist im Quellordner.
    private func quellVersion(_ eintrag: ProgrammEintrag) -> String {
        guard let projektDatei = eintrag.projektDatei, !projektDatei.istLeer else { return "" }
        let datei = (Pfade.repoWurzel as NSString).appendingPathComponent(projektDatei)
        guard FileManager.default.fileExists(atPath: datei) else { return "" }
        return Pfade.plistWert(datei, schluessel: "CFBundleShortVersionString")
    }
}
